### PR TITLE
feat(shorebird_code_push): add download update functionality

### PR DESCRIPTION
**Pull request by bryanoltman**
## Description

Adds support for triggering patch download from a Flutter app.

Also adds logging for ffi errors to help point users in the correct direction when things aren't working as expected.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: shorebird_code_push/lib/src/shorebird_code_push.dart [hunk 1]
```diff
@@ -2,48 +2,101 @@ import 'dart:isolate';
...
+      (updater) {
+        final patchNumber = updater.nextPatchNumber();
+        return patchNumber == 0 ? null : patchNumber;
+      },
+      fallbackValue: null,
+    );
+  }
+
+  /// Downloads the latest patch, if available.
+  Future<void> downloadUpdate() async {
```
Comment (felangel) [inline shorebird_code_push/lib/src/shorebird_code_push.dart:67]: Would be nice to return a Stream of progress updates eventually.

File: shorebird_code_push/example/lib/main.dart [hunk 2]
```diff
@@ -68,12 +70,26 @@ class _MyHomePageState extends State<MyHomePage> {
...
     );
   }
 
+  Future<void> _downloadUpdate() async {
+    setState(() {
+      _isDownloadingUpdate = true;
+    });
+
+    await _shorebirdCodePush.downloadUpdate();
+    _nextPatchVersion = await _shorebirdCodePush.nextPatchNumber();
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply shorebird_code_push/example/lib/main.dart:85]: > Can these ever throw?

Not as it stands. If the underlying ffi code throws, we catch those errors and log them.

File: shorebird_code_push/lib/src/shorebird_code_push.dart (hunk 1 above)
Comment (bryanoltman) [reply shorebird_code_push/lib/src/shorebird_code_push.dart:67]: https://github.com/shorebirdtech/updater/issues/38 :)

File: shorebird_code_push/lib/src/shorebird_code_push.dart [hunk 3]
```diff
@@ -2,48 +2,101 @@ import 'dart:isolate';
 
 import 'package:shorebird_code_push/src/updater.dart';
 
+/// A function that constructs an [Updater] instance. Used for testing.
+typedef UpdaterBuilder = Updater Function();
```
Comment (eseidel) [inline shorebird_code_push/lib/src/shorebird_code_push.dart:6]: You can use @visibleForTesting if you like, although that would add a package:meta dependency.

**Review by eseidel — APPROVED** [review]

**Review by felangel — APPROVED** [review]

File: shorebird_code_push/lib/src/shorebird_code_push.dart (hunk 3 above)
Comment (bryanoltman) [reply shorebird_code_push/lib/src/shorebird_code_push.dart]: We already had the meta dependency, so I did that and also made a `forTest` constructor so UpdaterBuilder isn't part of the public API.

**Pull request merged**